src/math_utils.rs

```diff
@@ -1,7 +1,29 @@
-use crate::bigint;
+use std::ops::Shr;
+
+use crate::{bigint, vm::errors::vm_errors::VirtualMachineError};
 use num_bigint::BigInt;
 use num_integer::Integer;
-use num_traits::{abs, FromPrimitive};
+use num_traits::{abs, FromPrimitive, Signed};
+
+///Returns the integer square root of the nonnegative integer n.
+///This is the floor of the exact square root of n.
+///Unlike math.sqrt(), this function doesn't have rounding error issues.
+pub fn isqrt(n: &BigInt) -> Result<BigInt, VirtualMachineError> {
+    //n.shr(1) = n.div_floor(2)
+    if !n.is_positive() {
```

Comment:
We probably want:
```suggestion
    if n.is_negative() {
```
`sqrt` is well defined for `0` (it's itself).

Reply:
Fixed!